src/Microsoft.AspNetCore.Routing/Matching/ILEmitTrieFactory.cs

```diff
@@ -0,0 +1,588 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#if IL_EMIT
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.AspNetCore.Routing.Matching
+{
+    internal static class ILEmitTrieFactory
+    {
+        // The algorthm we use only works for ASCII text. If we find non-ASCII text in the input
+        // we need to reject it and let is be processed with a fallback technique.
+        public const int NotAscii = Int32.MinValue;
+
+        public static Func<string, int, int, int> Create(
```

Comment:
Add a brief summary of what this is creating. It isn't immediately obvious from args and return type.